blockchain/blockchain.go

```diff
@@ -685,11 +686,33 @@ func (bc *blockchain) MintNewBlock(
 		return nil, err
 	}
 
-	// initial action iterator
-	actionIterator := actioniterator.NewActionIterator(actionMap)
-	actions, err := PickAction(genesis.BlockGasLimit, actionIterator)
-	// include coinbase transfer
-	actions = append(actions, selp)
+	// run execution and update state trie root hash
+	ws, err := bc.sf.NewWorkingSet()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to obtain working set from state factory")
+	}
+
+	gasLimitForContext := genesis.BlockGasLimit
+	ctx := protocol.WithRunActionsCtx(context.Background(),
+		protocol.RunActionsCtx{
+			BlockHeight:     newblockHeight,
+			BlockHash:       hash.ZeroHash32B,
+			ProducerPubKey:  producerPubKey,
+			BlockTimeStamp:  bc.now(),
+			ProducerAddr:    producerAddr,
+			GasLimit:        &gasLimitForContext,
+			EnableGasCharge: bc.config.Chain.EnableGasCharge,
+		})
+	root, rc, actions, err := bc.pickAndRunActions(ctx, selp, actionMap, ws)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to update state changes in new block %d", newblockHeight)
+	}
+
+	ra := block.NewRunnableActionsBuilder().
```

Comment:
NewRunnableActionsBuilder may not necessary to exist, but let's do it in another PR